internal/providers/manager/manager.go

```diff
@@ -104,9 +111,84 @@ func NewProviderManager(
 	return &providerManager{
 		classManagers: classes,
 		store:         store,
+		sessionStore:  sessionStore,
 	}, nil
 }
 
+func (p *providerManager) CreateFromConfig(ctx context.Context, providerClass db.ProviderClass, projectID uuid.UUID, name string, config json.RawMessage) (*db.Provider, error) {
+	manager, err := p.getClassManager(providerClass)
+	if err != nil {
+		return nil, fmt.Errorf("error getting class manager: %w", err)
+	}
+
+	provConfig, err := manager.GetConfig(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting provider config: %w", err)
+	}
+
+	// This method feels like it should be part of the class manager?
+	providerDef, err := providers.GetProviderClassDefinition(string(providerClass))
+	if err != nil {
+		return nil, fmt.Errorf("error getting provider definition: %w", err)
+	}
+
+	providerParams := db.CreateProviderParams{
+		Name:       name,
+		ProjectID:  projectID,
+		Class:      providerClass,
+		Implements: providerDef.Traits,
+		Definition: provConfig,
+		AuthFlows:  providerDef.AuthorizationFlows,
+	}
+
+	return p.store.Create(ctx, projectID, &providerParams)
+}
+
+func (p *providerManager) CreateFromSessionState(ctx context.Context, providerClass db.ProviderClass, creds v1.Credential, state string) (*db.Provider, error) {
+	manager, err := p.getClassManager(providerClass)
+	if err != nil {
+		return nil, fmt.Errorf("error getting class manager: %w", err)
+	}
+
+	stateData, err := p.sessionStore.GetBySessionState(ctx, state)
+	if err != nil {
+		return nil, fmt.Errorf("error getting state data by session state: %w", err)
+	}
+
+	// FIXME - this was in the old code
+	// p.mt.AddTokenOpCount(ctx, "check", stateData.RemoteUser.Valid)
+
+	encToken, err := manager.EncryptedCredentials(ctx, creds, state)
+	if err != nil {
+		return nil, fmt.Errorf("error encoding token: %w", err)
+	}
+
+	// This method feels like it should be part of the class manager?
+	providerDef, err := providers.GetProviderClassDefinition(string(providerClass))
+	if err != nil {
+		return nil, fmt.Errorf("error getting provider definition: %w", err)
+	}
+
+	providerParams := db.CreateProviderParams{
+		Name:       stateData.Provider,
+		ProjectID:  stateData.ProjectID,
+		Class:      providerClass,
+		Implements: providerDef.Traits,
+		Definition: json.RawMessage(stateData.ProviderConfig),
+		AuthFlows:  providerDef.AuthorizationFlows,
+	}
+
+	accessTokenParams := db.UpsertAccessTokenParams{
+		ProjectID:       stateData.ProjectID,
+		Provider:        stateData.Provider,
+		EncryptedToken:  encToken.EncodedData,
+		OwnerFilter:     stateData.OwnerFilter,
+		EnrollmentNonce: sql.NullString{String: state, Valid: true},
+	}
+
+	return p.store.CreateWithToken(ctx, stateData.ProjectID, &providerParams, &accessTokenParams)
```

Comment:
As discussed earlier, I like the idea of having the `ProviderManager` set up a token if we know it, but I feel like the current token table requires knowledge of things which don't belong in the `ProviderManager`, specifically enrolment nonce and owner filter.

I think we should leave the token management outside of this interface until we decouple those Github specific details out of the token table.